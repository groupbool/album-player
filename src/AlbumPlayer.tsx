import React from 'react';
import classNames from 'classnames';

import { Play, Pause, Spinner } from './utils/icons';
import AudioPlayer from './audioplayer';

interface Track {
	index: number;
	name: string;
	url: string;
}

interface AlbumPlayerProps {
	cover: string;
	title: string;
	tracks: Track[];
}

interface AlbumPlayerState {
	index: number;
	playing: boolean;
	busy: boolean;
	progress: number;
}

/**
 * Showcases an album art and lists all available track. Tracks
 * can be selected and playback can be controlled.
 */
export default class AlbumPlayer extends React.PureComponent<AlbumPlayerProps, AlbumPlayerState> {
	private player: AudioPlayer;

	constructor(props: AlbumPlayerProps) {
		super(props);
		this.state = {
			index: 0,
			playing: false,
			busy: false,
			progress: 0, // measure of completion
		};

		this.handleControl = this.handleControl.bind(this);
		this.handleTrackEnd = this.handleTrackEnd.bind(this);
		this.handleTrackSelect = this.handleTrackSelect.bind(this);
	}

	componentDidMount(): void {
		this.player = new AudioPlayer({
			onProgress: progress => this.setState({ progress }),
			onTrackEnd: this.handleTrackEnd,
		});
	}

	componentWillUnmount(): void {
		this.player.destroy();
	}

	handleTrackEnd(): void {
		const { tracks } = this.props;
		const { index } = this.state;
		const next = index + 1;
		if (next < tracks.length) {
			const track = tracks[next];
			console.debug('next track:', track.name);
			this.setState({ index: next, progress: 0, playing: false },
				() => this.handleControl(track.url));
		} else {
			this.setState({ playing: false });
		}
	}

	track(): Track {
		const { tracks } = this.props;
		const { index } = this.state;
		return tracks[index];
	}

	/**
	 * Handles play/pause behavior for the given audio url
	 * @param url The current audio url
	 */
	private handleControl(url: string) {
		this.setState({ busy: true }, () => {
			const { playing } = this.state;
			if (playing) {
				// pause track
				this.player.pause();
				this.setState({ busy: false, playing: false });
			} else {
				// play track
				this.player.play(url)
					.then(() => this.setState({ busy: false, playing: true }))
					.catch(err => {
						console.error('audio player error:', err);
						this.setState({ busy: false });
					});
			}
		});
	}

	handleTrackSelect(index: number): void {
		this.player.pause();
		this.setState({
			index: index,
			playing: false,
			progress: 0,
		});
	}

	render(): React.ReactNode {
		const { index, progress, playing, busy } = this.state;
		const { cover, title, tracks, } = this.props;
		const current = tracks[index];
		return (
			<div className="album">
				<div className="album-cover-container">
					<div className="album-controls-container">
						<TrackControls
							progress={progress}
							track={current}
							playing={playing}
							busy={busy}
							onToggle={() => this.handleControl(current.url)}
						/>
					</div>
					<img className="album-cover-img" src={cover} alt={title} />
				</div>
				<TrackList
					tracks={tracks}
					selected={index}
					onSelect={this.handleTrackSelect}
				/>
			</div>
		);
	}
}


/******************************************************************************/

interface ToggleButtonProps {
	playing: boolean;
	busy: boolean;
	onToggle: () => void;
}

function ToggleButton(props: ToggleButtonProps): JSX.Element {
	const { playing, busy } = props;
	if (busy) {
		return (
			<div className="media-controls-loading">
				<Spinner />
			</div>
		);
	}
	return (
		<i className="btn-control-container" onClick={props.onToggle}>
			{playing ? <Pause className="btn-control" /> : <Play className="btn-control" />}
		</i>
	);
}

/******************************************************************************/

interface TrackControlsProps {
	progress: number;
	track: Track;
	playing: boolean;
	busy: boolean;
	onToggle: () => void;
}

function TrackControls(props: TrackControlsProps): JSX.Element {
	const { progress, track } = props;
	return (
		<div className="album-controls">
			<div className="selected-info">
				{track.name}
			</div>
			<div className="media-controls">
				<ToggleButton
					playing={props.playing}
					busy={props.busy}
					onToggle={props.onToggle}
				/>
				<span className="track-progress">
					<span
						className="track-progress-current"
						style={{ width: `${progress}%` }}
					/>
				</span>
			</div>
		</div>
	);
}

/******************************************************************************/

interface TrackListProps {
	selected: number;
	tracks: Track[];
	onSelect: (index: number) => void;
}

function TrackList(props: TrackListProps): JSX.Element {
	const { selected, tracks } = props;
	return (
		<ol className="tracks">
			{tracks.map((track: Track, i: number) => {
				const trackClassName = classNames({
					'track': true,
					'track-selected': i == selected,
				});
				return (
					<li
						key={track.index}
						className={trackClassName}
						onClick={() => props.onSelect(i)}
					>
						<span className="track-number">{i + 1}</span> {track.name}
					</li>
				);
			})}
		</ol>
	);
}
