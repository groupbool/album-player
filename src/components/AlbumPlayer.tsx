import React from 'react';
import styled from 'styled-components'

import { Play, Pause, Spinner } from './icons';
import AudioPlayer from '../utils/audioplayer';

interface Track {
	index: number;
	name: string;
	url: string;
}

interface AlbumPlayerProps {
	cover: string;
	title: string;
	tracks: Track[];
	className?: string;
}

interface AlbumPlayerState {
	index: number;
	playing: boolean;
	busy: boolean;
	progress: number;
}

const AlbumPlayerContainer = styled.div`
	width: 100%;
	display: flex;
	flex-direction: column;
	background-color: #ecf0f1;
`;

const AlbumCoverContainer = styled.div`
	position: relative;
`;

const AlbumCoverImage = styled.img`
    display: block;
    width: 100%;
	height: auto;
`;

const AlbumControlsContainer = styled.div`
	position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    display: flex;
    align-items: flex-end;
`;

/**
 * Showcases an album art and lists all available track. Tracks
 * can be selected and playback can be controlled.
 */
export default class AlbumPlayer extends React.PureComponent<AlbumPlayerProps, AlbumPlayerState> {
	private player: AudioPlayer;

	constructor(props: AlbumPlayerProps) {
		super(props);
		this.state = {
			index: 0,
			playing: false,
			busy: false,
			progress: 0, // measure of completion
		};

		this.handleControl = this.handleControl.bind(this);
		this.handleTrackEnd = this.handleTrackEnd.bind(this);
		this.handleTrackSelect = this.handleTrackSelect.bind(this);
	}

	componentDidMount(): void {
		this.player = new AudioPlayer({
			onProgress: progress => this.setState({ progress }),
			onTrackEnd: this.handleTrackEnd,
		});
	}

	componentWillUnmount(): void {
		this.player.destroy();
	}

	handleTrackEnd(): void {
		const { tracks } = this.props;
		const { index } = this.state;
		const next = index + 1;
		if (next < tracks.length) {
			const track = tracks[next];
			console.debug('next track:', track.name);
			this.setState({ index: next, progress: 0, playing: false },
				() => this.handleControl(track.url));
		} else {
			this.setState({ playing: false });
		}
	}

	track(): Track {
		const { tracks } = this.props;
		const { index } = this.state;
		return tracks[index];
	}

	/**
	 * Handles play/pause behavior for the given audio url
	 * @param url The current audio url
	 */
	private handleControl(url: string) {
		this.setState({ busy: true }, () => {
			const { playing } = this.state;
			if (playing) {
				// pause track
				this.player.pause();
				this.setState({ busy: false, playing: false });
			} else {
				// play track
				this.player.play(url)
					.then(() => this.setState({ busy: false, playing: true }))
					.catch(err => {
						console.error('audio player error:', err);
						this.setState({ busy: false });
					});
			}
		});
	}

	handleTrackSelect(index: number): void {
		this.player.pause();
		this.setState({
			index: index,
			playing: false,
			progress: 0,
		});
	}

	render(): React.ReactNode {
		const { index, progress, playing, busy } = this.state;
		const { cover, title, tracks, className } = this.props;
		const current = tracks[index];
		return (
			<AlbumPlayerContainer className={className}>
				<AlbumCoverContainer>
					<AlbumControlsContainer>
						<TrackControls
							progress={progress}
							track={current}
							playing={playing}
							busy={busy}
							onToggle={() => this.handleControl(current.url)}
						/>
					</AlbumControlsContainer>
					<AlbumCoverImage src={cover} alt={title} />
				</AlbumCoverContainer>
				<TrackList
					tracks={tracks}
					selected={index}
					onSelect={this.handleTrackSelect}
				/>
			</AlbumPlayerContainer>
		);
	}
}


/******************************************************************************/

interface ToggleButtonProps {
	playing: boolean;
	busy: boolean;
	onToggle: () => void;
}

const Loading = styled.div`
	height: 1.5em;
    width: 1.5em;
`;

const ButtonContainer = styled.i`
	display: flex;
	align-items: center;
`;

function ToggleButton(props: ToggleButtonProps): JSX.Element {
	const { playing, busy } = props;
	if (busy) {
		return (
			<Loading>
				<Spinner />
			</Loading>
		);
	}
	return (
		<ButtonContainer onClick={props.onToggle}>
			{playing ? <Pause/> : <Play/>}
		</ButtonContainer>
	);
}

/******************************************************************************/

interface TrackControlsProps {
	progress: number;
	track: Track;
	playing: boolean;
	busy: boolean;
	onToggle: () => void;
}

const TrackControlsContainer = styled.div`
    background-color: rgba(52,73,94,.5);
    width: 100%;
    padding: 1em;
`;

const TrackControlsCurrent = styled.div`
    color: white;
	font-style: italic;
	height: 1.2em;
	font-size: 0.9em;
`;

const TrackMediaControls = styled.div`
	width: 100%;
	margin-top: 0.5em;
	display: flex;
	align-items: center;
`;

const TrackProgress = styled.div`
	background-color: white;
    height: 0.5em;
    flex-grow: 1;
    margin-left: 0.5em;
`;

const TrackCurrentProgress = styled.span<{progress: number}>`
	display: block;
	transition: 0.5s width;
	background-color: #b3b3b3;
	height: 0.5em;
	width: ${props => props.progress}%;
`;

function TrackControls(props: TrackControlsProps): JSX.Element {
	const { progress, track } = props;
	return (
		<TrackControlsContainer>
			<TrackControlsCurrent>
				{track.name}
			</TrackControlsCurrent>
			<TrackMediaControls>
				<ToggleButton
					playing={props.playing}
					busy={props.busy}
					onToggle={props.onToggle}
				/>
				<TrackProgress>
					<TrackCurrentProgress progress={progress}/>
				</TrackProgress>
			</TrackMediaControls>
		</TrackControlsContainer>
	);
}

/******************************************************************************/

interface TrackListProps {
	selected: number;
	tracks: Track[];
	onSelect: (index: number) => void;
}

const TrackListContainer = styled.ol`
	margin: 1em 0;
	padding: 0;
	list-style: none;
`;

const TrackListItem = styled.li<{selected: boolean}>`
	padding: 0.7em 1em;
	color: ${props => props.selected ? '#C86ED3' : 'inherit'};
	&:hover {
		cursor: pointer;
      	background-color: #dfe7e9;
	}
`;

const TrackIndex = styled.span`
	margin-right: 0.5em;
`;

function TrackList(props: TrackListProps): JSX.Element {
	const { selected, tracks } = props;
	return (
		<TrackListContainer>
			{tracks.map((track: Track, i: number) => {
				return (
					<TrackListItem
						key={track.index}
						selected={i == selected}
						onClick={() => props.onSelect(i)}
					>
						<TrackIndex>{i + 1}</TrackIndex> {track.name}
					</TrackListItem>
				);
			})}
		</TrackListContainer>
	);
}
