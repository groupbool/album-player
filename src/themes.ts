import { DefaultTheme } from 'styled-components';

export const BoolTheme: DefaultTheme = {
	font: 'monospace',
	base: 'black',
	controls: {
		base: 'white',
		background: 'rgba(52,73,94,.5)',
		hover: '#B3B3B3',
		fill: '#ccc',
	},
	tracks: {
		active: '#C86ED3',
		hover: '#dfe7e9',
	},
};