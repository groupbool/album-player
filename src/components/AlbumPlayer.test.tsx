import React from 'react';
import { shallow, mount } from 'enzyme';
import AlbumPlayer from './AlbumPlayer';
import { ThemeProvider } from 'styled-components';
import { BoolTheme } from '../themes';

// mocking HtmlMediaElement
window.HTMLMediaElement.prototype.load = () => { }; // eslint-disable-line
window.HTMLMediaElement.prototype.play = () => Promise.resolve();
window.HTMLMediaElement.prototype.pause = () => { }; // eslint-disable-line

const mixtape = {
	title: '2nd amendment',
	cover: 'https://f4.bcbits.com/img/a3659417884_16.jpg',
	artist: 'JPEGMAFIAxFreaky',
	tracks: [
		{
			url: 'https://jpegmafia.bandcamp.com/track/i-might-vote-4-donald-trump',
			name: 'I Might Vote 4 Donald Trump',
			index: 0,
		},
		{
			url: 'https://jpegmafia.bandcamp.com/track/trussmidaddi',
			name: 'Trussmidaddi',
			index: 1,
		},
	],
};

describe('rendering AlbumPlayer', () => {
	it('renders album player as expected', () => {
		const player = mount(
			<ThemeProvider theme={BoolTheme}>
				<AlbumPlayer {...mixtape}/>
			</ThemeProvider>
		);
		expect(player).toMatchSnapshot();
	});

	it('changes the current track on click', () => {
		const player = mount(
			<ThemeProvider theme={BoolTheme}>
				<AlbumPlayer {...mixtape}/>
			</ThemeProvider>
		);
		const tracks = player.find('li');
		tracks.at(1).simulate('click');
		const current = player.find({'data-tid': 'current-track'});
		expect(current.length).toBeGreaterThan(1);
		expect(current.at(0).text()).toEqual('Trussmidaddi');
	});
});
