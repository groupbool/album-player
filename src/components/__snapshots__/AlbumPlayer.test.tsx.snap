// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering AlbumPlayer renders album player as expected 1`] = `
<Component
  theme={
    Object {
      "base": "black",
      "controls": Object {
        "background": "rgba(52,73,94,.5)",
        "base": "white",
        "fill": "#ccc",
        "hover": "#B3B3B3",
      },
      "font": "monospace",
      "tracks": Object {
        "active": "#C86ED3",
        "hover": "#dfe7e9",
      },
    }
  }
>
  <AlbumPlayer
    artist="JPEGMAFIAxFreaky"
    cover="https://f4.bcbits.com/img/a3659417884_16.jpg"
    title="2nd amendment"
    tracks={
      Array [
        Object {
          "index": 0,
          "name": "I Might Vote 4 Donald Trump",
          "url": "https://jpegmafia.bandcamp.com/track/i-might-vote-4-donald-trump",
        },
        Object {
          "index": 1,
          "name": "Trussmidaddi",
          "url": "https://jpegmafia.bandcamp.com/track/trussmidaddi",
        },
      ]
    }
  >
    <styled.div>
      <div
        className="sc-crrsfI glheLf"
      >
        <styled.div>
          <div
            className="sc-dQppl cskZmc"
          >
            <styled.div>
              <div
                className="sc-kstrdz gQfiip"
              >
                <TrackControls
                  busy={false}
                  onToggle={[Function]}
                  playing={false}
                  progress={0}
                  track={
                    Object {
                      "index": 0,
                      "name": "I Might Vote 4 Donald Trump",
                      "url": "https://jpegmafia.bandcamp.com/track/i-might-vote-4-donald-trump",
                    }
                  }
                >
                  <styled.div>
                    <div
                      className="sc-fubCfw dBEmrO"
                    >
                      <styled.div
                        data-tid="current-track"
                      >
                        <div
                          className="sc-pFZIQ laZcVH"
                          data-tid="current-track"
                        >
                          I Might Vote 4 Donald Trump
                        </div>
                      </styled.div>
                      <styled.div>
                        <div
                          className="sc-jrAGrp chBoMP"
                        >
                          <ToggleButton
                            busy={false}
                            onToggle={[Function]}
                            playing={false}
                          >
                            <styled.i
                              onClick={[Function]}
                            >
                              <i
                                className="sc-iBPRYJ kMvjxY"
                                onClick={[Function]}
                              >
                                <Play>
                                  <styled.svg
                                    height="24"
                                    viewBox="0 0 24 24"
                                    width="24"
                                  >
                                    <svg
                                      className="sc-hKgILt fxcHoP"
                                      height="24"
                                      viewBox="0 0 24 24"
                                      width="24"
                                    >
                                      <path
                                        d="M0 0h24v24H0z"
                                        fill="none"
                                      />
                                      <path
                                        d="M10 16.5l6-4.5-6-4.5v9zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                                      />
                                    </svg>
                                  </styled.svg>
                                </Play>
                              </i>
                            </styled.i>
                          </ToggleButton>
                          <styled.div>
                            <div
                              className="sc-kEjbxe iisOsf"
                            >
                              <styled.span
                                progress={0}
                              >
                                <span
                                  className="sc-iqHYGH jKZUgz"
                                  style={
                                    Object {
                                      "width": "0%",
                                    }
                                  }
                                />
                              </styled.span>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                </TrackControls>
              </div>
            </styled.div>
            <styled.img
              alt="2nd amendment"
              src="https://f4.bcbits.com/img/a3659417884_16.jpg"
            >
              <img
                alt="2nd amendment"
                className="sc-bqyKva jBajkz"
                src="https://f4.bcbits.com/img/a3659417884_16.jpg"
              />
            </styled.img>
          </div>
        </styled.div>
        <TrackList
          onSelect={[Function]}
          selected={0}
          tracks={
            Array [
              Object {
                "index": 0,
                "name": "I Might Vote 4 Donald Trump",
                "url": "https://jpegmafia.bandcamp.com/track/i-might-vote-4-donald-trump",
              },
              Object {
                "index": 1,
                "name": "Trussmidaddi",
                "url": "https://jpegmafia.bandcamp.com/track/trussmidaddi",
              },
            ]
          }
        >
          <styled.ol>
            <ol
              className="sc-bdfBwQ ixPlzY"
            >
              <styled.li
                key="0"
                onClick={[Function]}
                selected={true}
              >
                <li
                  className="sc-gsTCUz cJBRwz"
                  onClick={[Function]}
                  selected={true}
                >
                  <styled.span>
                    <span
                      className="sc-dlfnbm gMYWRn"
                    >
                      1
                    </span>
                  </styled.span>
                   
                  I Might Vote 4 Donald Trump
                </li>
              </styled.li>
              <styled.li
                key="1"
                onClick={[Function]}
                selected={false}
              >
                <li
                  className="sc-gsTCUz dWYFlo"
                  onClick={[Function]}
                  selected={false}
                >
                  <styled.span>
                    <span
                      className="sc-dlfnbm gMYWRn"
                    >
                      2
                    </span>
                  </styled.span>
                   
                  Trussmidaddi
                </li>
              </styled.li>
            </ol>
          </styled.ol>
        </TrackList>
      </div>
    </styled.div>
  </AlbumPlayer>
</Component>
`;
